### PR TITLE
modules: mbedTLS: move debug log hook to modules/mbedtls/ and apply some improvements

### PR DESCRIPTION
**Pull request by mniestroj**
So far there was a debug log hook installed in TLS socket implementation.
However, mbedTLS (with debug enabled) might be used outside from TLS socket
and even outside from networking context.

Add new module, which implements debug log hook and makes it available
whenever CONFIG_MBEDTLS_DEBUG is enabled.

mbedTLS library threshold initialization was done in native TLS socket
implementation (which tends to use mbedTLS now) and inside mbedTLS
benchmark test. Move that to mbedTLS module initialization, as this is a
global setting.

Introduce a buildtime support of extracting basename of source filenames
logged using logging subsystem, so that there is no need to do it at
runtime.

Choose an appropriate Zephyr LOG_*() macro based on log level coming from
mbedTLS library. Remove log level number from formatted log messages, as it
is now redundant.

Debug log messages generated by mbedTLS library contain newline at the end
of log string. Remove this newline, if it exists, so that log output is
much more user friendly.

### CODE REVIEW

File: modules/mbedtls/Kconfig
Comment (rlubos) [inline modules/mbedtls/Kconfig:119]: Wouldn't it be better to make this a `choice` type configuration?

**Review by rlubos — COMMENTED** [review]
Very nice changes!

File: modules/mbedtls/Kconfig
Comment (mniestroj) [reply modules/mbedtls/Kconfig:123]: Indeed, `choice` sounds like a better option, as having both (buildtime and runtime) enabled at the same time doesn't seem to make sense. I've added 3rd option however to disable both, just in case somebody would want that.

**Review by rlubos — DISMISSED** [review]

**Review by ceolin — DISMISSED** [review]

**Review by d3zd3z — DISMISSED** [review]
I appreciate these changes. Thanks.

**Comment by mniestroj** [comment]
Rebased on `main`, so that fixes pulled in from `mcuboot` module have solved build failures. There was just minor conflict related to `log_strdup()` removal, which I solved (by removing `log_strdup()` use) as part of rebase.

**Comment by carlescufi** [comment]
@d3zd3z @ceolin @rlubos please review

**Review by rlubos — DISMISSED** [review]

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:131]: This is very unfortunate.

Someone building a project with a 3rd party toolchain will not see this message.
Such a user will just see (at best) a compilation error, but with no knowledge of why / how to fix.

We should not default to something which may not work generally.
That makes it much harder for users to know what to change, as well as require them to change configuration for any sample that uses mbed TLS.

Please ensure this setting is only default selection for toolchains that are known to work. (gcc)

**Review by tejlmand — CHANGES_REQUESTED** [review]
do not default to settings that we already know will break with some compilers.

File: modules/mbedtls/Kconfig
Comment (mniestroj) [reply modules/mbedtls/Kconfig:131]: >Such a user will just see (at best) a compilation error, but with no knowledge of why / how to fix.

Well, build error will be triggered for all Zephyr applications, as there is already a similar `zephyr_cc_option` in Zephyr's root `CMakeLists.txt`:

https://github.com/zephyrproject-rtos/zephyr/blob/2f5fef960e2310f5e4dd0bc1b9cd5fe9c5a73581/CMakeLists.txt#L401-L405

What I meant by:
```
	  This is compiler dependent, so if it does not work then please
	  fallback to MBEDTLS_DEBUG_EXTRACT_BASENAME_AT_RUNTIME instead.
```
is that it might not actually trim file paths. I assumed that this will not trigger build error, because we already do similar trick in root `CMakeLists.txt`.

> We should not default to something which may not work generally.

Agree on that, but what about root `CMakeLists.txt`?

> Please ensure this setting is only default selection for toolchains that are known to work. (gcc)

Any hint on how to achieve that?

File: subsys/net/lib/sockets/sockets_tls.c
Comment (plskeggs) [inline subsys/net/lib/sockets/sockets_tls.c:417]: This change prevents non-builtin MBed-TLS users from benefitting from these logging improvements.

If you changed it to:
`#if defined(MBEDTLS_DEBUG_C)`
then it will still work for both cases.

Comment (plskeggs) [reply subsys/net/lib/sockets/sockets_tls.c:417]: Any comments @mniestroj ?

Comment (mniestroj) [reply subsys/net/lib/sockets/sockets_tls.c:417]: > This change prevents non-builtin MBed-TLS users from benefitting from these logging improvements.

Yes, you are right. Initially I was trying to keep compatibility with "non-builtin" (`CONFIG_MBEDTLS_LIBRARY`). However the more I looked into that, the more existing issues I have found with that variant:
1. `modules/mbedtls/CMakeLists.txt` compiles in `zephyr_init.c` only for `CONFIG_MBEDTLS_BUILTIN=y`, which means that heap allocation is configured only for builtin variant,
2. `subsys/net/lib/sockets/sockets_tls.c` enables log messages with following code:
```
#if defined(MBEDTLS_DEBUG_C) && (CONFIG_NET_SOCKETS_LOG_LEVEL >= LOG_LEVEL_DBG)
	mbedtls_debug_set_threshold(CONFIG_MBEDTLS_DEBUG_LEVEL);
#endif
```
which in theory is enabled for external (non-builtin) mbedTLS variant, but if you look closer `CONFIG_MBEDTLS_DEBUG_LEVEL` is defined only when `CONFIG_MBEDTLS_DEBUG` is defined, which is only when `CONFIG_MBEDTLS_BUILTIN` is selected. So with `CONFIG_MBEDTLS_LIBRARY=y` Zephyr TLS socket glue layer wouldn't compile IMO (not sure, didn't try).
3. Any commits strictly related to `CONFIG_MBEDTLS_LIBRARY=y` are from few years back, so there is probably little or no users nowadays, which might be the reason why Zephyr TLS sockets would no longer build with external mbedTLS (see point 2.).
4. I am not aware of any build-test (or even a configuration outside of twister) of external mbedTLS library with Zephyr, so trying to setup this locally requires some more effort.

>If you changed it to:
>```
>#if defined(MBEDTLS_DEBUG_C)
>```
>then it will still work for both cases.

Unfortunately this is not enough. `CMakeLists.txt` needs to be updated to build "improved" `zephyr_mbedtls_debug()` function. As both: 1) heap setup and 2) setting mbedTLS log level, are not properly supported with `CONFIG_MBEDTLS_LIBRARY` (external mbedTLS), then I decided not to **try** to make it work.

@plskeggs Are you using external mbedTLS library actively with Zephyr?

File: modules/mbedtls/Kconfig
Comment (mniestroj) [reply modules/mbedtls/Kconfig:131]: @marc-hb Could you provide any input regarding `-fmacro-prefix-map` usage in this particular context? At least you are the author of 28a5657f1ff6 ("cmake: zephyr_cc_option(-fmacro-prefix-map=${ZEPHYR_BASE}=.)").

File: subsys/net/lib/sockets/sockets_tls.c
Comment (plskeggs) [reply subsys/net/lib/sockets/sockets_tls.c:417]: Yes, Nordic's nRF Connect SDK has its own fork of Mbed-TLS, with some helper code that takes care of the equivalent of `zephyr_init.c`.  I have a pending PR to that which adds calling `mbedtls_debug_set_threshold()` as needed.

I agree `CMakeLists.txt` would also need to change to compile in your new `zephyr_mbedtls_debug()` function.

Comment (mniestroj) [reply subsys/net/lib/sockets/sockets_tls.c:417]: I have a feeling that you are not using neither `CONFIG_MBEDTLS_BUILTIN` nor `CONFIG_MBEDTLS_LIBRARY`, but all the config symbols come from another Kconfig (not Zephyr upstream). Is that right? In that case I have really limited room for improvement, because the problem is not related to Zephyr upstream.

Comment (plskeggs) [reply subsys/net/lib/sockets/sockets_tls.c:417]: No, we use `CONFIG_MBEDTLS_LIBRARY`.

Comment (plskeggs) [reply subsys/net/lib/sockets/sockets_tls.c:417]: If you'd rather not accommodate this use case, I could put in a PR after yours merges to address my issues.

Comment (mniestroj) [reply subsys/net/lib/sockets/sockets_tls.c:417]: > No, we use CONFIG_MBEDTLS_LIBRARY.

Interesting. Could you provide me with example application that uses it, so I can give it a quick try?

>If you'd rather not accommodate this use case, I could put in a PR after yours merges to address my issues.

If you're okay with that, I could do my best to review such PR.

Comment (plskeggs) [reply subsys/net/lib/sockets/sockets_tls.c:417]: Sorry, my mistake.  It's not `CONFIG_MBEDTLS_LIBRARY` that we use, it's `CONFIG_MBEDTLS_TLS_LIBRARY` which is defined in one of our external libraries.

So, I'll wait for your PR to merge before I proceed with my own.

File: modules/mbedtls/Kconfig
Comment (tejlmand) [reply modules/mbedtls/Kconfig:131]: > > We should not default to something which may not work generally.
> 
> Agree on that, but what about root `CMakeLists.txt`?

There is a small difference to the CMakeLists.txt code and that is that the `zephyr_cc_option()` function will go through a compiler test, and if the compiler doesn't support the flag it will be discarded.

For example, it's not until [llvm 10](https://releases.llvm.org/10.0.0/tools/clang/docs/ReleaseNotes.html#new-compiler-flags) that this flag was supported.
So custom compilers based on older llvm (or gcc <v8.0.0) will not support the flag.
But those flags are not visible as a user config option.

When you expose the setting in Kconfig, then it gives the impression that this will work, thus misleading the user.
In fact the flag might be discarded on some compilers and such situations this can lead to confusion.
Why is my flag not being applied when enabled in Kconfig ?

> 
> > Please ensure this setting is only default selection for toolchains that are known to work. (gcc)
> 
> Any hint on how to achieve that?

Generally one can test for the compiler in Kconfig, for example like this:
```
config MBEDTLS_DEBUG_EXTRACT_BASENAME_AT_BUILDTIME
    ....
    depends on "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr"
```

because we know this toolchain definitely supports this setting.

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:149]: minor proposal, consider just this in title
```suggestion
 	bool "Strip newlines" 
```
we are already in a sub-menu related to mbed TLS debugging, so from where else would you be stripping newlines ?
Details are in the help text.

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:81]: it's a good thing you are now using the common log configuration template :+1:

But please remove the prompt text here: https://github.com/zephyrproject-rtos/zephyr/blob/30f07ed81f9f68d4c792ba0b062fc4d57a4a8589/modules/mbedtls/Kconfig#L98

else you'll end in the [stuck symbol syndrome](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig)

`ERROR` != `2`
image

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:134]: Just `runtime`, looks cleaner in menuconfig as selected choice in the
 `Extract basename from filenames (runtime)`
```suggestion
	bool "runtime"
```

File: modules/mbedtls/Kconfig
Comment (tejlmand) [inline modules/mbedtls/Kconfig:124]: Just `buildtime`, looks cleaner in menuconfig as selected choice in the
 `Extract basename from filenames (buildtime)`
```suggestion
	bool "buildtime"
```

**Review by tejlmand — CHANGES_REQUESTED** [review]
one blocker, the prompt in the log level conversion which suffers from stuck symbol.

And then some minor proposal, but not blockers.

File: modules/mbedtls/Kconfig
Comment (mniestroj) [reply modules/mbedtls/Kconfig:132]: Thanks for detailed explanation! I decided to enable "buildtime" only for zephyr toolchain, but still leave the option to select it manually for 3rdparty toolchains:
```
choice MBEDTLS_DEBUG_EXTRACT_BASENAME
	prompt "Extract basename from filenames"
	default MBEDTLS_DEBUG_EXTRACT_BASENAME_AT_BUILDTIME if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr"
	default MBEDTLS_DEBUG_EXTRACT_BASENAME_AT_RUNTIME
```

**Comment by mniestroj** [comment]
@tejlmand Thanks for detailed review! All of your comments should be addressed at this point, could you give it another look?

**Review by tejlmand — APPROVED** [review]
thanks for those improvements 👍

File: subsys/net/lib/sockets/sockets_tls.c
Comment (plskeggs) [reply subsys/net/lib/sockets/sockets_tls.c:417]: @tejlmand -- yes, that does work -- I tested it today with the commits from this PR.  However, without `CMakeLists.txt` changes, the new `debug.c` here does not get built.  And without `sockets_tls.c` changes, because we don't need `zephyr_init.c`, the call to `mbedtls_debug_set_threshold()` does not get made either.

I think on our side we should discuss either switching to CONFIG_MBEDTLS_LIBRARY, or define a new MBEDTLS_IMPLEMENTATION to cover our case, which would then build `debug.c` and call `mbedtls_debug_set_threshold()`.

**Review by jukkar — APPROVED** [review]

**Pull request merged**
